packages/grid/src/vaadin-grid-keyboard-navigation-mixin.js

```diff
@@ -920,26 +922,45 @@ export const KeyboardNavigationMixin = (superClass) =>
       focusTarget.tabIndex = isInteractingWithinActiveSection ? -1 : 0;
     }
 
-    /**
-     * @param {!HTMLTableRowElement} row
-     * @param {number} index
-     * @protected
-     */
-    _preventScrollerRotatingCellFocus(row, index) {
-      if (
-        row.index === this._focusedItemIndex &&
-        this.hasAttribute('navigating') &&
-        this._activeRowGroup === this.$.items
-      ) {
-        // Focused item has went, hide navigation mode
-        this._navigatingIsHidden = true;
-        this.toggleAttribute('navigating', false);
-      }
-      if (index === this._focusedItemIndex && this._navigatingIsHidden) {
-        // Focused item is back, restore navigation mode
-        this._navigatingIsHidden = false;
-        this.toggleAttribute('navigating', true);
+    /** @protected */
+    _preventScrollerRotatingCellFocus() {
+      if (this._activeRowGroup !== this.$.items) {
+        return;
       }
+
+      this.__preventScrollerRotatingCellFocusDebouncer = Debouncer.debounce(
+        this.__preventScrollerRotatingCellFocusDebouncer,
+        animationFrame,
+        () => {
+          const isItemsRowGroupActive = this._activeRowGroup === this.$.items;
+          const isFocusedItemRendered = this._getRenderedRows().some((row) => row.index === this._focusedItemIndex);
+          if (isFocusedItemRendered) {
+            // Ensure the correct element is focused, as the virtualizer
+            // may use different elements when re-rendering visible items.
+            this.__updateItemsFocusable();
```

Comment:
**note:** Debouncer + `__updateItemsFocusable` addresses https://github.com/vaadin/web-components/pull/7659#discussion_r1735671203. It also fixes cases where an incorrect item could be focused after scrolling back to the focused item triggered by <kbd>Tab</kbd> or an arrow key.